File: CPlus/longestArithSeqLength.cpp
class Solution {
public:
    int longestArithSeqLength(vector<int>& A) {
        unordered_map<int,unordered_map<int,int>> dp;
        if (A.size() <= 1) {
            return A.size();
        }
        int result=2;
        for (int i=0;i<A.size();i++){
            for (int j=0;j<i;j++){
                int diff=A[i]-A[j];
                if(!dp[diff].count(i)){
                    dp[diff][i]=2;
                }
                
                dp[diff][i]=max(dp[diff][j]+1,dp[diff][i]);
                result=max(result,dp[diff][i]);    
                
            }
        }
        return result;
    }
};

File: Python/longestArithSeqLength.py
class Solution(object):
    def longestArithSeqLength(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        if not A:
            return 0
        if len(A)<=2:
            return len(A)
        dp=[{} for i in range(len(A))]
        result=0
        
        for i in range(len(A)):
            for j in range(i):
                diff=A[i]-A[j]
                if diff in dp[j]:
                    dp[i][diff]=dp[j][diff]+1
                
                else:
                    dp[i][diff]=2
                
                result=max(result,dp[i][diff])
        
        
        return result